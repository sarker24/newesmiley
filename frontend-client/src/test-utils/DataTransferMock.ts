/*

This is sort of quick mock for data transfer, although it uses impl details (xhr).
Best case would be to directly mock the data transfer module api without having to deal with
impl details (or even move mock to core lib)

*/

const Methods = ['GET', 'POST', 'PATCH', 'PUT', 'DELETE'] as const;
type Method = typeof Methods[number];

type Options = {
  url: string;
  defaultStatus: number;
  query: boolean | ((queryParams: URLSearchParams) => boolean);
  response: MockResponseOptions;
  count?: number;
  always?: boolean;
};

// todo add headers
export type ResponseOptions =
  | string
  | [number, unknown]
  | {
      status?: number;
      response: unknown;
    };

export type RequestHeaders = { [header: string]: string };
export type RequestBody = string;

export type MockResponseOptions =
  | ResponseOptions
  | ((headers: RequestHeaders, body: RequestBody) => ResponseOptions);

const mockMap = (() => {
  let mocks: { [endpoint: string]: Options[] } = {};

  const add = (key: string, data: Options): void => {
    mocks[key] = [...(mocks[key] || []), data];
  };

  const get = (key: string): Options | undefined => {
    const mocksByKey = mocks[key];

    if (!mocksByKey) {
      return;
    }

    const [head, ...rest] = mocksByKey;

    if (head.always) {
      return head;
    }

    if (head.count === 1) {
      mocks[key] = rest.length > 0 ? rest : undefined;
      return head;
    }

    mocks[key] = [{ ...head, count: head.count - 1 }, ...rest];
    return head;
  };

  const reset = (): void => {
    mocks = {};
  };

  return {
    add,
    get,
    reset
  };
})();

const XMLHttpRequestMock = () => {
  let requestKey = null;
  const requestHeaders: RequestHeaders = {};
  let queryParams: URLSearchParams = null;

  const extractResponse = (body: any, responseOptions: Options) => {
    // todo headers + body check to mockMethod call args
    const response =
      typeof responseOptions.response === 'function'
        ? responseOptions.response(requestHeaders, body)
        : responseOptions.response;

    if (Array.isArray(response)) {
      return {
        status: response[0],
        response: response[1]
      };
    }

    const status = typeof response === 'string' ? responseOptions.defaultStatus : response.status;
    const data = typeof response === 'string' ? response : response.response;

    return { status, response: data };
  };

  const requestState = {
    onreadystatechange: () => {
      /* no op */
    },
    onerror: () => {
      /* no op */
    },
    open: (method: string, rawURL: string) => {
      const queryParamIndex = rawURL.indexOf('?');
      const url = queryParamIndex >= 0 ? rawURL.slice(0, queryParamIndex) : rawURL;
      queryParams =
        queryParamIndex >= 0
          ? new URLSearchParams(rawURL.slice(queryParamIndex))
          : new URLSearchParams('');
      requestKey = method + '_' + url;
    },
    send: (requestBody: any) => {
      const options = requestKey ? mockMap.get(requestKey) : undefined;
      if (
        options &&
        (typeof options.query === 'function' ? options.query(queryParams) : options.query)
      ) {
        const mockResponse = extractResponse(requestBody, options);
        requestState.readyState = 4;
        requestState.status = mockResponse.status;

        requestState.responseURL = options.url;
        requestState.responseText = JSON.stringify(mockResponse.response);
        requestState.response = JSON.stringify(mockResponse.response);
        requestState.onreadystatechange();
      } else {
        requestState.readyState = 4;
        requestState.status = 500;
        requestState.responseText = 'not mocked';
        requestState.response = 'not mocked';
        requestState.onreadystatechange();
      }
    },
    readyState: 0,
    setRequestHeader: (key: string, value: string) => (requestHeaders[key] = value),
    status: null,
    responseURL: null,
    response: null,
    responseText: null
  };

  return requestState;
};

// eslint-disable-next-line
// @ts-ignore
window.XMLHttpRequest = jest.fn().mockImplementation(XMLHttpRequestMock);

export default (baseURL?: string) => {
  const formattedBaseURL = baseURL.endsWith('/') ? baseURL.slice(0, -1) : baseURL;

  const buildURL = (path: string): string => {
    if (baseURL && !path.startsWith('http')) {
      const formattedPath = path.startsWith('/') ? path : '/' + path;
      return formattedBaseURL + formattedPath;
    }
    return path;
  };

  mockMap.reset();

  // todo: MockApiOptions = { query: boolean } | ((query: URLSearchParams) => boolean)
  type MockApiOptions = {
    query: boolean | ((queryParams: URLSearchParams) => boolean); // defaults to true, accepts any query params
  };

  type MockApiMethod = (url: string, options?: MockApiOptions) => ReturnType<typeof actions>;

  // TODO: onAny(*|url)
  type MockApi = {
    onGet: MockApiMethod;
    onPost: MockApiMethod;
    onPatch: MockApiMethod;
    onDelete: MockApiMethod;
    onPut: MockApiMethod;
  };

  const actions = (
    verb: Method,
    url: string,
    requestOptions: MockApiOptions = { query: true }
  ) => ({
    reply: (options: MockResponseOptions) => {
      const fullURL = buildURL(url);
      const key = verb + '_' + fullURL;
      mockMap.add(key, {
        query: requestOptions.query,
        defaultStatus: 200,
        response: options,
        url: fullURL,
        always: true
      });
      return mock;
    },
    replyOnce: (options: MockResponseOptions) => {
      const fullURL = buildURL(url);
      const key = verb + '_' + fullURL;
      mockMap.add(key, {
        query: requestOptions.query,
        defaultStatus: 200,
        response: options,
        url: fullURL,
        count: 1
      });
      return mock;
    },
    rejectOnce: (options: MockResponseOptions) => {
      const fullURL = buildURL(url);
      const key = verb + '_' + fullURL;
      mockMap.add(key, {
        query: requestOptions.query,
        defaultStatus: 500,
        response: options,
        url: fullURL,
        count: 1
      });
      return mock;
    },
    reject: (options: MockResponseOptions) => {
      const fullURL = buildURL(url);
      const key = verb + '_' + fullURL;
      mockMap.add(key, {
        query: requestOptions.query,
        defaultStatus: 500,
        response: options,
        url: fullURL,
        always: true
      });
      return mock;
    }
  });

  const mock = {
    reset: mockMap.reset,
    ...Methods.reduce<MockApi>(
      (all, m) => ({
        ...all,
        [`on${m[0].toUpperCase() + m.slice(1).toLowerCase()}`]: (url, options) =>
          actions(m, url, options)
      }),
      {} as MockApi
    )
  };

  return mock;
};
