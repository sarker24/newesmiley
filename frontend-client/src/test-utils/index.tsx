import React from 'react';
import { AnyAction, applyMiddleware, createStore } from 'redux';
import { Provider } from 'react-redux';
import rootReducer, { RootState } from 'redux/rootReducer';
import { mount as enzymeMount, shallow as enzymeShallow } from 'enzyme';
import { IntlProvider } from 'react-intl';
import { createMuiTheme, ThemeProvider } from '@material-ui/core/styles';
import palette from 'styles/palette';
import theme from 'styles/themes/global';
import ReduxThunk, { ThunkDispatch } from 'redux-thunk';
import ReduxPromise from 'redux-promise';
import { StylesProvider } from '@material-ui/core/styles';
import configureMockStore, { MockStoreEnhanced } from 'redux-mock-store';
import messages from '../i18n/en.json';

export { default as DataTransferMock } from './DataTransferMock';

const middleware = [ReduxThunk, ReduxPromise];
const initialRootState: RootState = rootReducer(undefined, {} as AnyAction);
// for consistent class names in testing
// eslint-disable-next-line
const generateClassName = (rule, styleSheet) => `${styleSheet.options.classNamePrefix}-${rule.key}`;

export type MountOptions = {
  initialState?: Partial<RootState>;
  [mountProp: string]: any;
};

export type ShallowOptions = MountOptions & { diveTo?: unknown };

function renderWithProviders(node: React.ReactElement, initialState: any = {}) {
  const store = createStore(rootReducer, initialState, applyMiddleware(...middleware));
  return (
    <Provider store={store}>
      <ThemeProvider theme={createMuiTheme({ ...theme, palette })}>
        <IntlProvider locale='en' defaultLocale='en' messages={messages}>
          <StylesProvider generateClassName={generateClassName}>{node}</StylesProvider>
        </IntlProvider>
      </ThemeProvider>
    </Provider>
  );
}

export function mount(node: React.ReactElement, options: MountOptions = {}) {
  const { initialState, ...mountOptions } = options;
  return enzymeMount(renderWithProviders(node, initialState), mountOptions);
}

export function shallow(node: React.ReactElement, options: ShallowOptions = {}) {
  const { initialState, diveTo, ...mountOptions } = options;
  const wrapper = enzymeShallow(renderWithProviders(node, initialState), mountOptions);

  if (!diveTo) {
    return wrapper;
  }

  let targetNode = wrapper;
  while (targetNode && targetNode.find(diveTo).length !== 1) {
    targetNode = targetNode.dive();
  }

  return targetNode ? targetNode.find(diveTo) : targetNode;
}

interface TypedMockStore
  extends MockStoreEnhanced<
    Partial<RootState>,
    ThunkDispatch<Partial<RootState>, void, AnyAction>
  > {
  getActions(): AnyAction[];
}

export function MockStore(state: Partial<RootState> = {}): TypedMockStore {
  const middlewares = [ReduxThunk, ReduxPromise];
  type DispatchExts = ThunkDispatch<Partial<RootState>, void, AnyAction>;
  return configureMockStore<Partial<RootState>, DispatchExts>(middlewares)({
    ...initialRootState,
    ...state
  });
}
