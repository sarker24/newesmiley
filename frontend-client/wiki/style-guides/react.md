# React/JSX Style Guide (WIP)

Inspired (ripped off) from: [Airbnb React/JSX Style Guide](https://github.com/airbnb/javascript/blob/master/react/README.md).

## Table of Contents

  1. [Basic guidelines](#basic-guidelines)
  1. [Alignment](#alignment)
  1. [Props](#props)
  1. [Component creation](#component-creation)
  1. [Naming](#naming)

## Basic guidelines

  - Optimize code for maintenance and readability before performance.
  - Include only one Component definition per file.
  - Divide your components into presentational and container components. [Read more](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0).

## Alignment

  - Follow these alignment styles for JSX syntax.

    ```jsx
    // bad
    <Foo superLongParam="bar"
         anotherSuperLongParam="baz" />

    // good
    <Foo
      superLongParam="bar"
      anotherSuperLongParam="baz"
    />

    // if props fit in one line then keep it on the same line
    <Foo bar="bar" />

    // children get indented normally
    <Foo
      superLongParam="bar"
      anotherSuperLongParam="baz"
    >
      <Quux />
    </Foo>
    ```
## Props

  - Always use camelCase for prop names.

    ```jsx
    // bad
    <Foo
      UserName="hello"
      phone_number={12345678}
    />

    // good
    <Foo
      userName="hello"
      phoneNumber={12345678}
    />
    ```

## Component creation

  - Always favor `functional` components over `class` components, unless you need to add [lifecycle hooks](https://facebook.github.io/react/docs/state-and-lifecycle.html#adding-lifecycle-methods-to-a-class) or other special cases:

    ```jsx
    // bad (this can be a function)
    class Listing extends React.Component {
      render() {
        return <div>{this.props.hello}</div>;
      }
    }

    // good
    function Listing (props) {
      return <div>{this.props.hello}</div>;
    };

    // good (using lifecycle methods)
    class Listing extends React.Component {
      componentDidMount() {
        this.props.onMount();
      }
      render() {
        return <div>{this.props.hello}</div>;
      }
    }
    ```

## Naming

  - **Extensions**: Use `.tsx` extension for React components.
  - **Filename**: Use PascalCase for filenames. E.g., `ReservationCard.jsx`.
  - **Reference Naming**: Use PascalCase for React components and camelCase for their instances.

    ```jsx
    // bad
    import reservationCard from './ReservationCard';

    // good
    import ReservationCard from './ReservationCard';

    // bad
    const ReservationItem = <ReservationCard />;

    // good
    const reservationItem = <ReservationCard />;
    ```

  - **Component Naming**: Use the filename as the component name. For example, `ReservationCard.jsx` should have a reference name of `ReservationCard`. However, for root components of a directory, use `index.tsx` as the filename and use the directory name as the component name:

    ```jsx
    // bad
    import Footer from './Footer/Footer';

    // bad
    import Footer from './Footer/index';

    // good
    import Footer from './Footer';
    ```

  - **Higher-order Component Naming**: Use a composite of the higher-order component's name and the passed-in component's name as the `displayName` on the generated component. For example, the higher-order component `withFoo()`, when passed a component `Bar` should produce a component with a `displayName` of `withFoo(Bar)`.

    > Why? A component's `displayName` may be used by developer tools or in error messages, and having a value that clearly expresses this relationship helps people understand what is happening.

    ```jsx
    // bad
    export default function withFoo(WrappedComponent) {
      return function WithFoo(props) {
        return <WrappedComponent {...props} foo />;
      }
    }

    // good
    export default function withFoo(WrappedComponent) {
      function WithFoo(props) {
        return <WrappedComponent {...props} foo />;
      }

      const wrappedComponentName = WrappedComponent.displayName
        || WrappedComponent.name
        || 'Component';

      WithFoo.displayName = `withFoo(${wrappedComponentName})`;
      return WithFoo;
    }
    ```

  - **Props Naming**: Avoid using DOM component prop names for different purposes.

    > Why? People expect props like `style` and `className` to mean one specific thing. Varying this API for a subset of your app makes the code less readable and less maintainable, and may cause bugs.

    ```jsx
    // bad
    <MyComponent style="fancy" />

    // good
    <MyComponent variant="fancy" />
    ```
