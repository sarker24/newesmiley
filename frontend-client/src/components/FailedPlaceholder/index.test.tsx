import * as React from 'react';
import { shallow } from 'test-utils';
import { FailedPlaceholder, default as Wrapper } from './index';

describe('FailedPlaceholder component', () => {
  it('should render with defaults', () => {
    const wrapper = shallow(<Wrapper />, { diveTo: FailedPlaceholder });

    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('should render with the button', () => {
    const wrapper = shallow(
      <Wrapper
        className='it-class'
        title='foo'
        description='bar'
        retryHandler={() => {
          /* no op*/
        }}
      />,
      { diveTo: FailedPlaceholder }
    );

    expect(wrapper.dive()).toMatchSnapshot();
  });
});
