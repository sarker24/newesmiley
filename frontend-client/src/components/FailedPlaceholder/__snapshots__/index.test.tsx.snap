// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FailedPlaceholder component should render with defaults 1`] = `
<div
  className="placeholder FailedPlaceholder-container-1"
>
  <SVGInline
    classSuffix="-svg"
    cleanup={Array []}
    cleanupExceptions={Array []}
    component="span"
    height="125px"
    svg="test-file-stub"
    width="100px"
  />
  <WithStyles(ForwardRef(Typography))
    gutterBottom={true}
    variant="body1"
  >
    Whoops!
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))
    align="center"
    gutterBottom={true}
    variant="subtitle1"
  >
    Something went wrong while communicating with us.
  </WithStyles(ForwardRef(Typography))>
</div>
`;

exports[`FailedPlaceholder component should render with the button 1`] = `
<div
  className="placeholder FailedPlaceholder-container-1 it-class"
>
  <SVGInline
    classSuffix="-svg"
    cleanup={Array []}
    cleanupExceptions={Array []}
    component="span"
    height="125px"
    svg="test-file-stub"
    width="100px"
  />
  <WithStyles(ForwardRef(Typography))
    gutterBottom={true}
    variant="body1"
  >
    foo
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))
    align="center"
    gutterBottom={true}
    variant="subtitle1"
  >
    bar
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Button))
    onClick={[Function]}
    startIcon={<Memo />}
  >
    Retry
  </WithStyles(ForwardRef(Button))>
</div>
`;
